dali/python/nvidia/dali/ops/_signatures.py

```diff
@@ -131,15 +131,87 @@ def _get_positional_input_param(schema, idx):
                          annotation=annotation)
 
 
-def _get_positional_input_params(schema):
+def _get_annotation_input_regular():
+    """Return the annotation for regular input parameter in DALI, used for the primary overload.
+    A function is used as a global variable can be confused with type alias.
+    TODO(klecki): Extend with TensorLike.
+    """
+    return _DataNode
+
+
+
+def _get_annotation_return_regular(schema):
+    """Produce the return annotation for DALI operator suitable for primary, non-MIS overload.
+    Note the flattening, single output is not packed in Sequence.
+    """
+    if schema.HasOutputFn():
+        # Dynamic number of outputs, not known at "compile time"
+        return_annotation = Union[_DataNode, Sequence[_DataNode], None]
+    else:
+        # Call it with a dummy spec, as we don't have Output function
+        num_regular_output = schema.CalculateOutputs(_b.OpSpec(""))
+        if num_regular_output == 0:
+            return_annotation = None
+        elif num_regular_output == 1:
+            return_annotation = _DataNode
+        else:
+            # Here we could utilize the fact, that the tuple has known length, but we can't
+            # as DALI operators return a list
+            # Also, we don't advertise the actual List type, hence the Sequence.
+            return_annotation = Sequence[_DataNode]
+    return return_annotation
+
+
+def _get_annotation_input_mis():
+    """Return the annotation for multiple input sets, used for the secondary operator overload.
+    A function is used as a global variable can be confused with type alias.
+    """
+    return Union[_DataNode, List[_DataNode]]
+
+
+def _get_annotation_return_mis(schema):
+    """Annotation for function that handles Multiple input sets overload.
+    Note that DALI does a lot of flattening, so single-element sequences are transformed to
+    just the element.
+    """
+    if schema.HasOutputFn():
+        # Dynamic number of outputs, not known at "compile time"
+        # We can return single or multiple outputs in regular case (same as primary overload),
+        # a list of single outputs or a list of multiple outputs for MIS or None.
+        return_annotation = Union[_DataNode, Sequence[_DataNode], List[_DataNode],
+                                  List[Sequence[_DataNode]], None]
+    else:
+        # Call it with a dummy spec, as we don't have Output function
+        num_regular_output = schema.CalculateOutputs(_b.OpSpec(""))
+        if num_regular_output == 0:
+            return_annotation = None
+        elif num_regular_output == 1:
+            # This allows for type hints with single-return operators to work correctly with MIS
+            # as the return type matches the overload for their input type.
+            return_annotation = Union[_DataNode, List[_DataNode]]
```

Comment:
MIS overload never returns a single DataNode.

Reply:
`fn.copy([input])` returns a DataNode as opposed to `fn.copy([input, input])` which returns `List[DataNode]`.